src/Functions/FunctionBinaryArithmetic.h

```diff
@@ -1667,31 +1673,51 @@ class FunctionBinaryArithmetic : public IFunction
                 {
                     if constexpr (IsDataTypeDecimal<LeftDataType> && IsDataTypeDecimal<RightDataType>)
                     {
-                        if constexpr (is_division)
+                        if constexpr (is_div_int || is_div_int_or_zero)
+                            type_res = std::make_shared<ResultDataType>();
```

Comment:
I think we can move this condition one layer outside:
```c++
if constexpr (is_div_int || is_div_int_or_zero)
{
    type_res = std::make_shared<ResultDataType>();
}
else if constexpr (IsDataTypeDecimal<LeftDataType> && IsDataTypeDecimal<RightDataType>)
{
    ...
}
```
So we won't need to check `is_div_int || is_div_int_or_zero` in any other scopes